### ISSUE TITLE
Memory Leak on route navigation

### ISSUE THREAD

**Issue by warflash** [nuxt/framework#8414]
### Environment

------------------------------
- Operating System: `Linux`
- Node Version:     `v16.14.2`
- Nuxt Version:     `3.0.0-rc.12`
- Nitro Version:    `0.6.0`
- Package Manager:  `npm@7.17.0`
- Builder:          `vite`
- User Config:      `-`
- Runtime Modules:  `-`
- Build Modules:    `-`
------------------------------

### Reproduction

https://v3.nuxtjs.org/getting-started/introduction

### Describe the bug

repeatedly switch tabs from e.g. `get started` to `guide` and watch the memory increase.
Fresh load
![image](https://user-images.githubusercontent.com/22006082/197361818-10e11af2-e8c4-4ca9-b356-05e40ae8eeda.png)
After 30 swaps
![image](https://user-images.githubusercontent.com/22006082/197361829-1373a5da-fb7e-4f99-8b79-4ae9a1d37504.png)

### Additional context

For bigger pages this results in 100s of leaked MBs within only a handful of route changes which in turn makes switching routes take seconds and is really bad user experience.

**Comment by warflash** [comment]
@danielroe Just to make sure, this is not limited to the docs (since you added the docs label) and rather seems like common behaviour

**Comment by danielroe** [comment]
@warflash Can you identify any commonalities between your site and the docs? (Are you using Nuxt content, for example?) Are you able to provide a minimal reproduction?

**Comment by warflash** [comment]
Hey @danielroe we are using content indeed, so I tried removing it and the issue persisted. 
I do not have a minimal reproduction but the same behaviour can be see with nuxt movies which you guys recently released, which is somewhat minimal I guess? 
https://movies.nuxt.space/search
swapping back and fourth between any page takes up ram which even when force garbage collecting is not given back. It doesnt grow as quickly as in our app here, probably due to the app being simpler but the behaviour is the exact same.
Initial load:
![image](https://user-images.githubusercontent.com/22006082/197385357-cccf8c5a-dd18-4aaa-9f89-86e38bf04cac.png)
20 swaps or so between `/search` and `/tv`, force gc'ed afterwards:
![image](https://user-images.githubusercontent.com/22006082/197385382-1c8a0a26-801c-4926-8937-ac483f43755b.png)
Seems to be even worse on firefox, gotten up to 60mb on movies with just a few navigations.

**Comment by warflash** [comment]
Oh and same issue on [nuxt hackernews](https://github.com/nuxt/hackernews), which again, not the most minimal but just goes to show it seems like a really common issue. Swap between https://hn.nuxt.space/news/1 and https://hn.nuxt.space/newest/1 and slowly see the memory rise and keep allocated after GC's. Also verified same nuxt HN behaviour on firefox

**Comment by danielroe** [comment]
Would you try adding the following to your `nuxt.config` and letting me know the result?

```js
export default defineNuxtConfig({
  app: {
    pageTransition: false,
    layoutTransition: false,
  },
})
```

**Comment by warflash** [comment]
Unfortunately that doesnt seem to help, also tried with a local clone of nuxt movies and didn't reduce the issue there either

**Comment by danielroe** [comment]
@warflash Are you sure? I've been trying with Nuxt Movies locally and it greatly reduces the issue. (Note that you'll almost certainly need to build + run it in production mode to confirm.)

**Comment by warflash** [comment]
Oh ok, I just tested in dev mode as the behaviour there was the same as in prod. Will recheck with a built project

**Comment by danielroe** [comment]
I often regret interim updates, but - 🤷 - my current hypothesis is that this is an issue with `<Suspense>` + `<Transition>` preserving the context in the pending branch, meaning every click handler (for example) preserves a reference to potentially thousands of removed DOM nodes.

Having entirely disabled `<NuxtPage>` and `<NuxtLayout>` in the Nuxt Movies project, I still get the behaviour with the following template. (If you want to replicate, you will also need to swap out the auto-imports of `useRoute` and `useRouter` to be from `vue-router` for as pure a Vue experience as possible).

```vue
<template>
  <div>
    <RouterView v-slot="{ Component }">
      <template v-if="Component">
        <Transition mode="out-in">
          <Suspense>
            <component :is="Component" />
          </Suspense>
        </Transition>
      </template>
    </RouterView>
  </div>
</template>
```

https://vuejs.org/guide/built-ins/suspense.html#combining-with-other-components

Removing `<Transition>`, either from the template above, or if using `NuxtPage`, via `app.pageTransition` and `app.layoutTransition`  _greatly_ reduces the issue in my testing (that is, in the Nuxt Movies repro, decreases leak from 1.7Mb per navigation to 70Kb). Nevertheless, there are still some preserved detached DOM nodes from the pending state of `<Suspense>` , which suggests that the real core issue here is coming from `<Suspense>`.

The next step is likely to attempt to reproduce using Vue without Nuxt to confirm my hypothesis.

Possible causes I've ruled out:
- ~~NuxtLink's intersection observer implementation~~
- ~~NuxtLayout~~
- ~~NuxtPage and page progress hooks~~
- ~~useHead~~

**Comment by warflash** [comment]
Yep you were correct, that does fix the memory accumulation of repeated navigations indeed. There still seems to be a bit of weirdness left though. Visiting different routes (not the same ones over and over) does also make the memory go up and never deallocates.
![image](https://user-images.githubusercontent.com/22006082/197396996-86ec1824-78f9-4153-ac65-672dfe67c999.png)
Not sure if the cause for that is the same (as in: any navigation allocates memory that's never given back, setting page and layout transition to false prevents repeated allocations for the same route but on different routes you still end up with the same behaviour)

**Comment by danielroe** [comment]
Baby's up from his nap, so I'll have to come back to this later. If you get a chance, would you try to replicate with https://github.com/nuxt-contrib/vue3-ssr-starter? It'll be tricky as it's so much smaller a project, but 🤷‍♂️

**Comment by warflash** [comment]
Gotcha, got a pretty tight schedule today unfortunately but should I get to it I'll post it here. Enjoy family time and the rest of your sunday!
